File: dynamic_array/dynamic_array.py
class DynamicArray:
    def __init__(self, capacity: int):
        if capacity <= 0:
            raise ValueError("Capacity must be greater than 0.")
        self.capacity = capacity
        self.size = 0
        self.array = [None] * capacity

    def get_size(self) -> int:
        return self.size

    def is_empty(self) -> bool:
        return self.size == 0

    def get(self, index: int):
        if index < 0 or index >= self.size:
            raise IndexError("Index out of bounds.")
        return self.array[index]

    def set(self, index: int, value):
        if index < 0 or index >= self.size:
            raise IndexError("Index out of bounds.")
        self.array[index] = value

    def add(self, value):
        if self.size == self.capacity:
            self._resize()
        self.array[self.size] = value
        self.size += 1

    def _resize(self):
        new_capacity = self.capacity * 2
        new_array = [None] * new_capacity
        for i in range(self.size):
            new_array[i] = self.array[i]
        self.array = new_array
        self.capacity = new_capacity

    def remove(self, index: int):
        if index < 0 or index >= self.size:
            raise IndexError("Index out of bounds.")
        for i in range(index, self.size - 1):
            self.array[i] = self.array[i + 1]
        self.array[self.size - 1] = None
        self.size -= 1  

    def pop(self):
        if self.is_empty():
            raise IndexError("Cannot pop from an empty array.")
        value = self.array[self.size - 1]
        self.array[self.size - 1] = None
        self.size -= 1
        return value

    def __str__(self):
        return str(self.array[:self.size])

# Example usage:
if __name__ == "__main__":
    arr = DynamicArray(2)
    arr.add(1)
    arr.add(2)
    print(arr) 

    arr.add(3)
    print(arr)

    arr.remove(1)
    print(arr)

    print(arr.get(0))
    arr.set(0, 10)
    print(arr)

    print(arr.pop())
    print(arr)
    print(arr.is_empty())
    arr.pop()
    print(arr.is_empty())

    

File: dynamic_array/dynamic_array.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int* array;
    int capacity;
    int size;
} DynamicArray;

void initArray(DynamicArray* arr, int initialCapacity) {
    if (initialCapacity <= 0) {
        printf("Initial capacity must be greater than 0.\n");
        exit(1);
    }
    arr->capacity = initialCapacity;
    arr->size = 0;
    arr->array = (int*)malloc(initialCapacity * sizeof(int));
}

void freeArray(DynamicArray* arr) {
    free(arr->array);
    arr->array = NULL;
    arr->capacity = 0;
    arr->size = 0;
}

int getSize(DynamicArray* arr) {
    return arr->size;
}

int getCapacity(DynamicArray* arr) {
    return arr->capacity;
}

int get(DynamicArray* arr, int index) {
    if (index < 0 || index >= arr->size) {
        printf("Index out of bounds.\n");
        exit(1);
    }
    return arr->array[index];
}

int set(DynamicArray* arr, int index, int value) {
    if (index < 0 || index >= arr->size) {
        printf("Index out of bounds.\n");
        exit(1);
    }
    arr->array[index] = value;
}

void resize(DynamicArray* arr) {
    arr->capacity *= 2;
    int* newArray = (int*)malloc(arr->capacity * sizeof(int));
    for (int i = 0; i < arr->size; i++) {
        newArray[i] = arr->array[i];
    }
    free(arr->array);
    arr->array = newArray;
}

void add(DynamicArray* arr, int value) {
    if (arr->size == arr->capacity) {
        resize(arr);
    }
    arr->array[arr->size++] = value;
}

void removeAt(DynamicArray* arr, int index) {
    if (index < 0 || index >= arr->size) {
        printf("Index out of bounds.\n");
        exit(1);
    }
    for (int i = index; i < arr->size - 1; i++) {
        arr->array[i] = arr->array[i + 1];
    }
    arr->size--;
}

int pop(DynamicArray* arr) {
    if (arr->size == 0) {
        printf("Array is empty.\n");
        exit(1);
    }
    return arr->array[--arr->size];
}

int main() {
    DynamicArray arr;
    initArray(&arr, 2);

    add(&arr, 10);
    add(&arr, 20);
    add(&arr, 30);

    for (int i = 0; i < getSize(&arr); i++) {
        printf("%d ", get(&arr, i));
    }
    printf("\n");

    pop(&arr);
    for (int i = 0; i < getSize(&arr); i++) {
        printf("%d ", get(&arr, i));
    }
    printf("\n");

    removeAt(&arr, 0);
    for (int i = 0; i < getSize(&arr); i++) {
        printf("%d ", get(&arr, i));                    
    }
    printf("\n");

    freeArray(&arr);
    return 0;
}